numpy/lib/mixins.py

```diff
@@ -0,0 +1,167 @@
+"""Mixin classes for custom array types that don't inherit from ndarray."""
+from __future__ import division, absolute_import, print_function
+
+import sys
+
+from numpy.core import umath as um
+
+# None of this module should be exposed in top-level NumPy module.
+__all__ = []
+
+
+def _binary_method(ufunc):
+    def func(self, other):
+        try:
+            if other.__array_ufunc__ is None:
+                return NotImplemented
+        except AttributeError:
+            pass
+        return self.__array_ufunc__(ufunc, '__call__', self, other)
```

Comment:
@njsmith, @shoyer - my comment above was specifically on what implications `__array_ufunc__ = None` should have for other classes and dispatch. I think there should be only implications for the class itself: that *it* cannot handle `ufunc`, i.e., that it is equivalent to returning `NotImplemented`. I don't see a reason to prohibit a priory that other classes can deal with it inside a `ufunc`.